components/org.wso2.carbon.healthcheck.api.core/src/main/java/org/wso2/carbon/healthcheck/api/core/impl/DataSourceHealthChecker.java

```diff
@@ -0,0 +1,318 @@
+        Integer maxActiveCount = getNumber(maxActive);
+
+        validateAvailableResourceMargin(activeCount, maxActiveCount, datasouceName, errors);
+        return properties;
+    }
+
+    protected Properties validateDataSource(DataSource dataSource, String datasourceName, List<HealthCheckError>
+            errors) {
+
+        Properties properties = new Properties();
+
+        if (dataSource instanceof DataSourceProxy) {
+            DataSourceProxy dataSourceProxy = (DataSourceProxy) dataSource;
+            ConnectionPool pool = dataSourceProxy.getPool();
+            int activeCount = pool.getActive();
+            int maxActive = dataSourceProxy.getMaxActive();
+            properties.put(datasourceName + ACTIVE_CONNECTION_COUNT_KEY, activeCount);
+            validateAvailableResourceMargin(activeCount, maxActive, datasourceName, errors);
+        }
+
+        return properties;
+    }
+
+    protected void validateAvailableResourceMargin(Integer activeCount, Integer maxActiveCount,
+                                                   String datasorceName, List<HealthCheckError> errors) {
+
+        if (activeCount != null && maxActiveCount != null) {
+            if (activeCount > (maxActiveCount * poolHealthyPercentage / 100)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Datasource: " + datasorceName + " surpasses the healthy connection percentage of " +
+                            poolHealthyPercentage + ". Current usage is: " + activeCount);
+                }
+                errors.add(new HealthCheckError(Constants.ErrorCodes.ERROR_CONNECTION_COUNT_EXCEEDS_LIMIT,
+                        "Active count of db connections exceeds the " +
+                                "given limit. Active count: " + activeCount + " for datasource: " + datasorceName));
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Data source: " + datasorceName + " has a healthy pool usage of : " + activeCount);
+                }
+            }
+        }
+    }
+
+    protected Object getMbeanAttribute(String objectName, String attributeName) throws HealthCheckFailedException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Retrieving attribute: " + attributeName + ", From Mbean: " + objectName);
+        }
+
+        MBeanServer mbs = ManagementFactory.getPlatformMBeanServer();
+        ObjectName mbeanName;
+        try {
+            mbeanName = new ObjectName(objectName);
+        } catch (MalformedObjectNameException e) {
+            throw new HealthCheckFailedException("No MBeans are registered with object name: " + objectName, e);
+        }
+        try {
+            return mbs.getAttribute(mbeanName, attributeName);
+        } catch (MBeanException | AttributeNotFoundException | InstanceNotFoundException | ReflectionException e) {
+            throw new HealthCheckFailedException("Error while reading attribute: " + attributeName + ", " +
```

Comment:
JMX is not enabled for data sources by default. So it would be better to give a clue in error message to check whether JMX is enabled.